### PR TITLE
Enable IPython display

### LINKED ISSUE #38: ipython notebook html display for jobs
**[Original report](https://bitbucket.org/glotzer/signac/issue/38) by Anonymous.**

----------------------------------------

Since ipython/jupyter notebook can display html output, I thought it might be nice to display jobs in a table (the same way pandas dataframes are displayed). So I copied an [online example](http://calebmadrigal.com/display-list-as-table-in-ipython-notebook/):

```
#!python

class JobTable(list):
    """ Overridden list class which takes a list of signac jobs 
        and renders an HTML table in IPython Notebook with the id
        and statepoint attributes. """
    
    def _repr_html_(self):
        """Return html table representing myself, the list of jobs.

        Operations
        ----------
        1. Determine what the attributes are.
        2. Add header row with name of each attribute.
        3. Add row to table for each job, with job id in first column.
        """
        attribs = set(sum((job.statepoint().keys() for job in self), []))
        html = ["<table>"]
        html.append("<td><b>{0}</b></td>".format('id'))
        for attrib in attribs:
            html.append("<td><b>{0}</b></td>".format(attrib))
            
        for job in self:
            html.append("<tr>")
            html.append("<td><b>{0}</b></td>".format(job.get_id()))
            
            for attrib in attribs:
                try:
                    html.append("<td>{0}</td>".format(job.statepoint()[attrib]))
                except KeyError:
                    html.append("<td></td>")
            
            html.append("</tr>")
        html.append("</table>")
        return ''.join(html)
```

This can then be used:

```
#!python
from IPython.display import display
display(JobTable(project.find_jobs({'type':'pmf'})))
```

And an HTML table describing the jobs appears in the results cell (image attached). If this were implemented as part of signac, it would be cleaner to change `find_jobs` from a generator function to instead return an iterator class with the `_repr_html_` method. Do you the pretty notebook display is worth embedding into signac?

### PR DESCRIPTION
**Pull request by bdice**
## Description
Resolves #38 by adding IPython display capabilities for the `JobsCursor` class.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

This pull request is based on
- [ ] *master*, because it is a bug fix or an update to the documentation.
- [x] *develop*, because it introduces a new feature.

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.txt).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt).

### CODE REVIEW

**Comment by bdice** [comment]
@glotzerlab/signac-developers I'd like some input on the desired appearance of this PR, specifically wrt nested keys. I was going to show them as dot-separated. The current code probably doesn't work for nested keys. Also how should this be tested? 

image

Note: job 99914b932bd37a50b983c5e7c90ae93b has statepoint `{}`.

**Comment by csadorf** [comment]
I'd flatten all keys, so treat all keys separately and join nested keys with dots. 

There should be some fixtures for various valid statepoint (basic and nested, valid types) in the job unit tests. I'd make sure that this function doesn't fail for any of those and compare the raw output for a few select cases.

**Comment by bdice** [comment]
Another question: how many rows should we show? It's probably good to clip it.

Pandas has configurable package-wide settings for this. Default is 60 rows and 8 columns. An alternative would be to just return the pandas dataframe and allow its `_repr_html_` method to be called. That introduces a requirement for pandas to have pretty-printed HTML, but it would give users the ability to configure the display the same way they would already do for pandas dataframes. It also simplifies our display and testing logic by a lot.

https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

```python
pd.options.display.chop_threshold      pd.options.display.max_colwidth
pd.options.display.colheader_justify   pd.options.display.max_info_columns
pd.options.display.column_space        pd.options.display.max_info_rows
pd.options.display.date_dayfirst       pd.options.display.max_rows
pd.options.display.date_yearfirst      pd.options.display.max_seq_items
pd.options.display.encoding            pd.options.display.memory_usage
pd.options.display.expand_frame_repr   pd.options.display.multi_sparse
pd.options.display.float_format        pd.options.display.notebook_repr_html
pd.options.display.html                pd.options.display.pprint_nest_depth
pd.options.display.large_repr          pd.options.display.precision
pd.options.display.latex               pd.options.display.show_dimensions
pd.options.display.max_categories      pd.options.display.unicode
pd.options.display.max_columns         pd.options.display.width
```

**Comment by csadorf** [comment]
I like the idea of delgating to pandas. How would we deal with pandas not being installed?

**Comment by bdice** [comment]
> 
> 
> I like the idea of delgating to pandas. How would we deal with pandas not being installed?

I would just return `str(self)` or so, similar to the existing default behavior.

**Comment by bdice** [comment]
## With 🐼:
image

## Without 🐼:
image


(commit 6b5c5e7cf9a064a148a5bcfb77b893051e89e7c0)

**Comment by bdice** [comment]
Should this also be the default Jupyter Notebook representation for `Project` or `Job`, possibly with some alteration?

**Comment by csadorf** [comment]
> Should this also be the default Jupyter Notebook representation for `Project` or `Job`, possibly with some alteration?

For the `Project` it would be important to also show the directories I think. What would the `Job` HTML view look like?

File: signac/contrib/project.py [hunk 1]
```diff
@@ -1827,6 +1827,15 @@ def _export_sp_and_doc(job):
             data={job._id: dict(_export_sp_and_doc(job)) for job in self},
             orient='index').infer_objects()
 
+    def _repr_html_(self):
+        """Represent a JobsCursor in IPython Notebooks."""
+        try:
+            # Attempt to use pandas representation
+            return self.to_dataframe()._repr_html_()
+        except ImportError:
+            import warnings
```
Comment (csadorf) [inline signac/contrib/project.py]: Can you move that import to the top of the module?

File: changelog.txt [hunk 2]
```diff
@@ -29,6 +29,7 @@ Added
  - Add automatic cast of numpy arrays to lists when storing them within a `JSONDict`, e.g., a `job.statepoint` or `job.document`.
  - Enable `Collection` class to manage collections stored in gzip files.
  - Enable deleting of `JSONDict` keys through the attribute interface, e.g., `del job.doc.foo`.
+ - Added HTML representation of JobsCursor objects for Jupyter Notebooks (uses pandas).
```
Comment (csadorf) [inline changelog.txt]: Please use present tense and code-markup for `JobsCursor`.

**Pull request merged**